source/lac/trilinos_sparse_matrix.cc

```diff
@@ -126,9 +126,6 @@ namespace TrilinosWrappers
           return;
         }
 
-      // otherwise first flush Trilinos caches
-      matrix->compress ();
-
```

Comment:
OK, this is actually an interesting case. The code we previously had was actively wrong in that if you looped over all of the locally owned rows, you'd call `compress()` for each of these rows. But that means calling `compress()` a different number of times on each processor. This may have worked by accident if individual processors (wrongly) decide to elide the communication if nothing had been changed locally, but it would fail as soon as one processor has no rows at all.

Furthermore, it is extremely inefficient. If you loop over the locally owned elements of a parallel matrix (or, just all elements of a non-parallel matrix), then this calls `compress()` many times even though it is likely not going to make any kind of difference.

In other words, I think the call here was inserted because it seemed like the safe thing to do, but not because it was right. That said, I'm going to add a notice to the `begin()` functions where we create these iterators stating that the matrix needs to be "clean" before using iterators on it.